table/tables/state_remote.go

```diff
@@ -0,0 +1,275 @@
+	LockForRead(tid int64, now, ts uint64) (bool, error)
+
+	// LockForWrite try to add a write lock to the table with the specified tableID
+	LockForWrite(tid int64, now, ts uint64) error
+
+	// RenewLease attempt to renew the read lock on the table with the specified tableID
+	RenewLease(tid int64, ts uint64) (bool, error)
+}
+
+// mockStateRemoteHandle implement the StateRemote interface.
+type mockStateRemoteHandle struct {
+	ch chan remoteTask
+}
+
+func (r *mockStateRemoteHandle) Load(tid int64) (CachedTableLockType, uint64, error) {
+	op := &loadOP{tid: tid}
+	op.Add(1)
+	r.ch <- op
+	op.Wait()
+	return op.lockType, op.lease, op.err
+}
+
+func (r *mockStateRemoteHandle) LockForRead(tid int64, now, ts uint64) (bool, error) {
+	op := &lockForReadOP{tid: tid, now: now, ts: ts}
+	op.Add(1)
+	r.ch <- op
+	op.Wait()
+	return op.succ, op.err
+}
+
+func (r *mockStateRemoteHandle) LockForWrite(tid int64, now, ts uint64) error {
+	op := &lockForWriteOP{tid: tid, now: now, ts: ts}
+	op.Add(1)
+	r.ch <- op
+	op.Wait()
+	if op.err != nil {
+		return errors.Trace(op.err)
+	}
+	// No block, finish.
+	if op.oldLease == 0 {
+		return nil
+	}
+
+	// Wait for read lock to expire.
+	t1 := oracle.GetTimeFromTS(op.oldLease)
+	t2 := oracle.GetTimeFromTS(now)
+	waitDuration := t1.Sub(t2)
+	time.Sleep(waitDuration)
+
+	// TODO: now should be a new ts
+	op = &lockForWriteOP{tid: tid, now: op.oldLease + 1, ts: leaseFromTS(op.oldLease + 1)}
+	op.Add(1)
+	r.ch <- op
+	op.Wait()
+	// op.oldLease should be 0 this time.
+	return op.err
+}
+
+func (r *mockStateRemoteHandle) RenewLease(tid int64, ts uint64) (bool, error) {
+	return false, errors.New("not implemented yet")
+}
+
+func mockRemoteService(r *mockStateRemoteData, ch chan remoteTask) {
+	for task := range ch {
+		task.Exec(r)
+	}
+}
+
+type remoteTask interface {
+	Exec(data *mockStateRemoteData)
+}
+
+// loadOP is a kind of remoteTask
+type loadOP struct {
+	sync.WaitGroup
+	// Input
+	tid int64
+
+	// Output
+	lockType CachedTableLockType
+	lease    uint64
+	err      error
+}
+
+func (op *loadOP) Exec(data *mockStateRemoteData) {
+	op.lockType, op.lease, op.err = data.Load(op.tid)
+	op.Done()
+}
+
+// lockForReadOP is a kind of rmoteTask
+type lockForReadOP struct {
+	sync.WaitGroup
+	// Input
+	tid int64
+	now uint64
+	ts  uint64
+
+	// Output
+	succ bool
+	err  error
+}
+
+func (op *lockForReadOP) Exec(r *mockStateRemoteData) {
+	op.succ, op.err = r.LockForRead(op.tid, op.now, op.ts)
+	op.Done()
+}
+
+// lockForWriteOP is a kind of remote task
+type lockForWriteOP struct {
+	sync.WaitGroup
+	// Input
+	tid int64
+	now uint64
+	ts  uint64
+
+	// Output
+	err      error
+	oldLease uint64
+}
+
+func (op *lockForWriteOP) Exec(data *mockStateRemoteData) {
+	op.oldLease, op.err = data.LockForWrite(op.tid, op.now, op.ts)
+	op.Done()
+}
+
+type mockStateRemoteData struct {
+	data map[int64]*stateRecord
+}
+
+type stateRecord struct {
+	lease         uint64
+	readLockLease uint64 // only use for intent lock
```

Comment:
The meaning of `lease` and `readLockLease` is so confusing. Why not introduce two leases as `lockLease` and `writeIntentLease`  ?

Reply:
you are right.